Chezhanov ivan mp2 lab1 set

IceHoles opened the pull request.

IceHoles committed:
Laba commit

IceHoles committed:
add all funcs

IceHoles committed:
a lot of fixes

IceHoles commented:
@PanovaElena
Долго разбирался с гитом, но вроде как разобрался, написал функции, зафиксил чтобы проходили все тесты

PanovaElena commented on src/tbitfield.cpp:
```diff
@@ -7,78 +7,134 @@
+    if (n < 0 || n > BitLen) {
+        throw std::invalid_argument("Negative bit");
+    }
+    return pMem[GetMemIndex(n)] & GetMemMask(n);;
```
Я предполагала, что тут вы вернете 0 или 1.

PanovaElena commented:
Все работает. Один вопрос:

Пусть есть такой тест, который на вашей реализации красный:

```
TEST(TBitField, can_do_AND_correctly) {
    const int size1 = 70, size2 = 35;
    TBitField bf1(size1), bf2(size2), bf3(size1);
    for (int i = 0; i < size1; i += 2)
        bf1.SetBit(i);
    for (int i = 1; i < size2; i += 2)
        bf2.SetBit(i);

    TBitField res1 = bf1 & bf2;
    TBitField res2 = bf2 & bf1;

    EXPECT_EQ(bf3, res1);
    EXPECT_EQ(bf3, res2);
}
```

Должен ли он работать? Корректен ли он? Или проблема в реализации? Вопрос неоднозначный...

IceHoles committed:
Added can_do_AND_correctly test
changed GetBit function so it returns actual bit instead of a number
changed OR operator and AND operator logic in order to make them work properly

IceHoles commented:
@PanovaElena изменил функцию GetBit чтобы вертала не число а значение бита и поменял операторы чтобы работали по новой логике, в таком случае при операции над двумя битфилдами разных размеров будет возвращаться битфилд большего размера и как если бы он сравнивался с меньшим, дополненным ноликами до размера большего

PanovaElena commented:
Принято
